### PR TITLE
Fix some issues with the xds-protos package

### PR DESCRIPTION
**Pull request by lidizheng**
Last PR https://github.com/grpc/grpc/pull/25975 failed to discover several important issues, because the test was way too weak. A pre-release package has been uploaded to https://pypi.org/project/xds-protos/. Manually tested. I will upload a new release after this PR.

Fixes
- `0.0.1` failed to include non-envoy protos
- Make `xds-protos` compatible with Google's namespace packages
- Split the build process from `setup.py` (`setup.py` will always be imported during the install process)

Improvements
- Added a README
- Make sure the test is truly isolated

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/25999)

### CODE REVIEW

**Comment by lidizheng** [comment]
@gnossen PTAL.

File: tools/distrib/python/xds_protos/build.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    '--proto_path={}'.format(VALIDATE_ROOT),
+    '--proto_path={}'.format(WELL_KNOWN_PROTOS_INCLUDE),
+    '--proto_path={}'.format(OPENCENSUS_PROTO_ROOT),
+    '--python_out={}'.format(OUTPUT_PATH),
+]
+COMPILE_BOTH = COMPILE_PROTO_ONLY + ['--grpc_python_out={}'.format(OUTPUT_PATH)]
+
+
+def has_grpc_service(proto_package_path: str) -> bool:
+    return proto_package_path.startswith('envoy/service')
```
Comment (gnossen) [inline tools/distrib/python/xds_protos/build.py:72]: Is this assumption guaranteed going forward? What's the behavior when you use `--grpc_python_out` on a file with no services? If it emits an empty file or doesn't emit any file at all, you might just want to unconditionally supply this flag.

File: tools/distrib/python/xds_protos/build.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            if file_name.endswith('.proto'):
+                # Compile proto
+                if has_grpc_service(proto_package_path):
+                    return_code = protoc.main(COMPILE_BOTH +
+                                              [os.path.join(root, file_name)])
+                    add_test_import(proto_package_path, file_name, service=True)
+                else:
+                    return_code = protoc.main(COMPILE_PROTO_ONLY +
+                                              [os.path.join(root, file_name)])
+                    add_test_import(proto_package_path,
```
Comment (gnossen) [inline tools/distrib/python/xds_protos/build.py:91]: This is a really interesting approach to testing this. Where do you run the generated test file?

**Review by gnossen — APPROVED** [review]

File: tools/distrib/python/xds_protos/build.py (hunk 1 above)
Comment (lidizheng) [reply tools/distrib/python/xds_protos/build.py:72]: Our ProtoBuf plugin will create a file if the `--grpc_python_out` flag is given. The file will have a line to import the generated `_pb.py` of that proto. When trying to import generated proto modules, the extra `_pb_grpc` will appear in the code hinter (I'm using VSCode). So, I added the logic to only generate service when we are certain there is a gRPC service.

So far, I think Envoy protos are doing a good job to keep services under `envoy/service`. If there are any exceptions, I can update the code. Or, do you think we should use regex to read the proto file?

File: tools/distrib/python/xds_protos/build.py (hunk 2 above)
Comment (lidizheng) [reply tools/distrib/python/xds_protos/build.py:91]: The test script will be generated at `tools/distrib/python/xds_protos` when the `build.py` script is executed. The test itself will be run in `tools/distrib/python/xds_protos/build_validate_upload.sh`, before we uploading the package to PyPI.

File: tools/distrib/python/xds_protos/build.py (hunk 1 above)
Comment (gnossen) [reply tools/distrib/python/xds_protos/build.py:72]: > So far, I think Envoy protos are doing a good job to keep services under envoy/service

If there is a problem, it should happen infrequently enough that manually fixes should suffice.

File: tools/distrib/python/xds_protos/build.py (hunk 2 above)
Comment (gnossen) [reply tools/distrib/python/xds_protos/build.py:91]: > The test itself will be run in tools/distrib/python/xds_protos/build_validate_upload.sh

Ah, there we go. Missed that bit.

**Comment by lidizheng** [comment]
Known failures: https://github.com/grpc/grpc/issues/25242 https://github.com/grpc/grpc/issues/26022 https://github.com/grpc/grpc/issues/26023

**Pull request merged**